python/semantic_kernel/agents/orchestration/magentic.py

```diff
@@ -0,0 +1,854 @@
+    ] = ChatHistory()
+    participant_descriptions: Annotated[
+        dict[str, str], Field(description="The descriptions of the participants in the group.")
+    ]
+    round_count: Annotated[int, Field(description="The number of rounds completed.")] = 0
+    stall_count: Annotated[int, Field(description="The number of stalls detected.")] = 0
+    reset_count: Annotated[int, Field(description="The number of resets detected.")] = 0
+
+    def reset(self) -> None:
+        """Reset the context.
+
+        This will clear the chat history and reset the stall count.
+        This won't reset the task, round count, or participant descriptions.
+        """
+        self.chat_history.clear()
+        self.stall_count = 0
+        self.reset_count += 1
+
+
+# endregion Messages and Types
+
+# region MagenticManager
+
+
+@experimental
+class MagenticManager(KernelBaseModel, ABC):
+    """Base class for the Magentic One manager."""
+
+    max_stall_count: Annotated[int, Field(description="The maximum number of stalls allowed before a reset.", ge=0)] = 3
+    max_reset_count: Annotated[int | None, Field(description="The maximum number of resets allowed.", ge=0)] = None
+    max_round_count: Annotated[
+        int | None, Field(description="The maximum number of rounds (agent responses) allowed.", gt=0)
+    ] = None
+
+    @abstractmethod
+    async def plan(self, magentic_context: MagenticContext) -> ChatMessageContent:
+        """Create a plan for the task.
+
+        This is called when the task is first started.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ChatMessageContent: The task ledger.
+        """
+        ...
+
+    @abstractmethod
+    async def replan(self, magentic_context: MagenticContext) -> ChatMessageContent:
+        """Replan for the task.
+
+        This is called when the task is stalled or looping.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ChatMessageContent: The updated task ledger.
+        """
+        ...
+
+    @abstractmethod
+    async def create_progress_ledger(self, magentic_context: MagenticContext) -> ProgressLedger:
+        """Create a progress ledger.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ProgressLedger: The progress ledger.
+        """
+        ...
+
+    @abstractmethod
+    async def prepare_final_answer(self, magentic_context: MagenticContext) -> ChatMessageContent:
+        """Prepare the final answer.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ChatMessageContent: The final answer.
+        """
+        ...
+
+
+@experimental
+class StandardMagenticManager(MagenticManager):
```

Comment:
```suggestion
class MagenticManager(MagenticManagerBase):
```
I think it would be better if people can clearly see by name that the base class doesn't work out of the box and that the dev experience is better if they can just start with the MagenticManager without having to customize a bunch of stuff!

Reply:
I think we are keeping the `StandardMagenticManager` as it requires structured output. We have purposely made it in the base class so that devs can extend it without the structured output requirement.